sound/soc/sof/intel/hda-pcm.c

```diff
@@ -95,6 +95,83 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
 	}
 };
 
+/* There are two dma modes in dsp:
+ * (1) host dma and link dma work in decouple mode for SOF + HDA
+ * (2) host dma and GP dma work in couple mode for SOF + I2S
+ * The difference is that in decopule mode, the host dma is set by dai
+ * frontend and link dma is set by dai backend and the setting maybe
+ * different while in couple mode, host dma and GP dma share the same
+ * dma setting.
+ */
+static inline int get_be_dma_channel(struct snd_sof_dev *sdev,
+				     struct snd_pcm_substream *substream,
+				     struct sof_ipc_stream_params *ipc_params)
+{
+	struct snd_soc_pcm_runtime *fe = substream->private_data;
+	struct hdac_ext_stream *be_stream = NULL;
+	struct snd_soc_dpcm *dpcm;
+	int direction = substream->stream;
+	struct sof_intel_be_params *be_params;
+
+	be_params = (struct sof_intel_be_params *)ipc_params->private_data;
+	be_params->be_count = 0;
+
+	/* travel BE clients to get BE stream with link dma id */
+	list_for_each_entry(dpcm, &fe->dpcm[direction].be_clients, list_be) {
+		struct snd_soc_pcm_runtime *be = dpcm->be;
+		struct snd_soc_dapm_widget *dapm_widget;
+		struct snd_sof_widget *sof_widget;
+		int index = be_params->be_count;
+
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+			dapm_widget = be->cpu_dai->playback_widget;
+		else
+			dapm_widget = be->cpu_dai->capture_widget;
+
+		sof_widget = snd_sof_find_swidget(sdev, dapm_widget->name);
+		if (!sof_widget) {
+			dev_err(sdev->dev, "error: failed to find backend widget");
+			return -EINVAL;
+		}
+
+		be_stream = snd_soc_dai_get_dma_data(be->cpu_dai, substream);
+		if (be_stream) {
```

Comment:
Can you clarify what this test means.

Reply:
The be_stream is stored by HDA dai backend. For I2S backend, there is no such operation.
So check it to make clear the backend is HDA or I2S